php-checks/src/main/java/org/sonar/php/checks/CountInsteadOfEmptyCheck.java

```diff
@@ -89,6 +92,99 @@ private boolean isCountFunction(FunctionCallTree tree) {
     return FUNCTION_PREDICATE.test(TreeValues.of(tree, context().symbolTable()));
   }
 
+  private static final Tree.Kind[] USED_STATEMENTS = {
+    Tree.Kind.EXPRESSION_STATEMENT,
+    Tree.Kind.RETURN_STATEMENT,
+    Tree.Kind.IF_STATEMENT,
+    Tree.Kind.WHILE_STATEMENT,
+    Tree.Kind.DO_WHILE_STATEMENT,
+    Tree.Kind.FOR_STATEMENT,
+    Tree.Kind.FOREACH_STATEMENT,
+    Tree.Kind.SWITCH_STATEMENT,
+    Tree.Kind.ASSIGNMENT,
+    Tree.Kind.FUNCTION_DECLARATION,
+    Tree.Kind.METHOD_DECLARATION};
+
+  private static final Tree.Kind[] AND_OR_TREE = {Tree.Kind.CONDITIONAL_AND, Tree.Kind.CONDITIONAL_OR, Tree.Kind.ALTERNATIVE_CONDITIONAL_AND, Tree.Kind.ALTERNATIVE_CONDITIONAL_OR};
+
+  // Checks if empty() is used on the same variable in the surrounding expression context
+  // Returns true if empty() call is found, indicating we should skip raising an issue
+  private boolean isEmptyUsedInExpression(FunctionCallTree countCallTree, @Nullable ExpressionTree countArgument) {
```

Comment:
I suggest to talk about conditions in this comment and in the function's name since expressions can be almost anything (e.g., `a + 42`).
Furthermore, I think this comment should be a JavaDoc comment and could use some refinement. Here is my suggestion:
```suggestion
  /**
   * Checks if "empty" is called for the given variable in the current condition.
   * For example, in case of "empty($foo) && count($foo) > 0" we want to know if "empty" is called in addition to "count".
   * @param countCallTree represents the count call in the current condition.
   * @param countArgument the argument passed to the call of count.
   * @return true if there exists an "empty" call for the countArgument in the scope of the current condition.
   */
  private boolean isEmptyUsedInCondition(FunctionCallTree countCallTree, @Nullable ExpressionTree countArgument) {
```

Reply:
Still not fully taken into account 😉